alphagov/paas-cf
# [#155463715] Move global manifest properties to specific jobs

[#155463715 Move global manifest properties to specific jobs](https://www.pivotaltracker.com/story/show/155463715)

What?
----

In order to move closer to the manifests from [cf-deployment](https://github.com/cloudfoundry/cf-deployment) we want to move all the global properties
to each specific job in the manifest.


How to review
-------------

This change is meant to result in no real changes in the deployment,
a No-Op.

**To test the deployment**:

 1. Deploy from master (be sure this branch is rebased) 
 2. Deploy this branch
    - bosh would display a lot of changes in the properties diff, but then it would be a no-op, no VM would be updated at all.
    - all test should pass

**Verify the new manifest**

We have created a set of tool scripts in the repo https://github.com/keymon/paas-cf-release-reference/tree/master/scripts

You can check that the new manifest is equivalent by using the script  https://github.com/keymon/paas-cf-release-reference/blob/master/scripts/manifest_to_specific_properties.rb. This script, given a rendered cf-manifest.yml and a directory with all the releases spec (e.g. `/tmp/releases/capi-release/jobs/*/spec`) would generate a manifest with the properties in each job.

To use it:

```
git clone https://github.com/keymon/paas-cf-release-reference
cd paas-cf-release-reference
```

Generate the manifest based on master, and expand the variables into each specific job using the spec_job:

```
PAAS_CF_PATH=~/gds/paas-cf

( cd $PAAS_CF_PATH && git checkout master )

spruce merge \
    $PAAS_CF_PATH/manifests/cf-manifest/manifest/*.yml \
    <($PAAS_CF_PATH/manifests/cf-manifest/scripts/grafana-dashboards-manifest.rb $PAAS_CF_PATH/manifests/cf-manifest/grafana/) \
    <(ruby $PAAS_CF_PATH/terraform/scripts/generate_vpc_peering_manifest.rb "$PAAS_CF_PATH/terraform/prod.vpc_peering.json"
) \
    < $PAAS_CF_PATH/manifests/cf-manifest/stubs/datadog-nozzle.yml \
    < $PAAS_CF_PATH/manifests/cf-manifest/stubs/oauth.yml \
    > spruced-manifest.yml

RELEASES_DIR=./releases-spec/ ./scripts/manifest_to_specific_properties.rb ./spruced-manifest.yml > expanded-spruced-manifest.yml

```

The file `expanded-spruced-manifest.yml` will be our reference.

Now generate the manifest based on current brach:

```
PAAS_CF_PATH=~/gds/paas-cf

( cd $PAAS_CF_PATH && git checkout move_properties_to_specific_jobs_155463715 )

spruce merge \
    $PAAS_CF_PATH/manifests/cf-manifest/manifest/*.yml \
    <($PAAS_CF_PATH/manifests/cf-manifest/scripts/grafana-dashboards-manifest.rb $PAAS_CF_PATH/manifests/cf-manifest/grafana/) \
    <(ruby $PAAS_CF_PATH/terraform/scripts/generate_vpc_peering_manifest.rb "$PAAS_CF_PATH/terraform/prod.vpc_peering.json"
) \
    < $PAAS_CF_PATH/manifests/cf-manifest/stubs/datadog-nozzle.yml \
    < $PAAS_CF_PATH/manifests/cf-manifest/stubs/oauth.yml \
    > spruced-manifest-v2.yml


```

Finally diff both manifests with spruce diff. There should not be any difference.

```
spruce diff \
 <(spruce merge --prune meta expanded-spruced-manifest.yml) \
 <(spruce merge --prune meta spruced-manifest-v2.yml)  | less -R

```

**Verify removed properties**:

You can use the script https://github.com/keymon/paas-cf-release-reference/blob/master/scripts/print_all_properties_in_manifest.rb to print the properties that are defined somewhere in a manifest:

```
./scripts/print_all_properties_in_manifest.rb spruced-manifest.yml > spruce-manifest.properties.txt
./scripts/print_all_properties_in_manifest.rb spruced-manifest-v2.yml > spruce-manifest-v2.properties.txt


diff spruce-manifest.properties.txt spruce-manifest-v2.properties.txt
```


Who?
----

Anyone but @keymon or @LeePorte
